IntrusiveEdgesSpecifics: reduce Map lookups and prevent invalid edge reuse

HannesWell opened the pull request:
With this PR I suggest to improve `UniformIntrusiveEdgesSpecifics` and `WeightedIntrusiveEdgesSpecifics` in the following ways:

- Reduce the number of map lookups from two to one (i.e. the call to `Map.contains()` respectively to `Map.put()` is skipped).
- Detect if an `IntrusiveEdge` is already contained to another graph but with different touching  vertices and reject the edge in this case.
Previously the `source` and `target` field were just overridden. This led to an inconsistent state of the graph data-structure.
Alternatively an exception could be thrown to indicate the issue.

In `AbstractBaseGraph.addEdge(V sourceVertex, V targetVertex, E e)` in the case of a graph that does not support multiple-edges, the `intrusiveEdgesSpecifics.containsEdge(e)`-check could also be removed. It is just a performance penalty in most cases. It only saves runtime in case the edge is already present in this or another graph but with other touching vertices. And I think this is a quite rare case compared to just add a really new edge or adding an edge that is already in the graph and connects the two specified vertices (the later case is detected by `specifics.addEdgeToTouchingVerticesIfAbsent(sourceVertex, targetVertex, e)`).

<!-- describe the changes you have made here: what, why, …
     Link issues by using the following pattern: [#333](https://github.com/jgrapht/jgrapht/issues/333).
     The title of the PR must not reference an issue, because GitHub does not support autolinking there. -->

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [ ] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

jsichi commented:
Could you merge from master?  We had a problem with Travis CI, so pull requests weren't getting CI build runs.

jsichi commented:
I think throwing an exception would be best for this case, let's call it IntrusiveEdgeException, and provide the existing source/target as context.  @d-michail do you agree?

d-michail commented:
Yes!

HannesWell commented:
I changed it accordingly. If an attempt is made to add an `IntrusiveEdge` that is already associated with other vertices, an `IntrusiveEdgeException` is now thrown.
This required special handling in the AbstractBaseGraph because in case an Exception is thrown the edge added to the `specifics` needs to be removed again (to keep the graph in an consistent state). I unified this handling for both `addEdge` methods.

Additionally I unified the adding of IntrusiveEdges and the handling.

Furthermore I removed the check `intrusiveEdgesSpecifics.containsEdge` in `addEdge(V, V, E)`.
In most cases this check is just a performance penalty. And because the rejection of an edge from the EdgeSpecific has to be handled anyway (since it can throw an exception) there is no need to ensure up front that adding will succeed. An up-front check only saves runtime in case the edge is already present in this or another graph but with other touching vertices (and is not an intrusive edge). Compared to other cases (for example if just a valid edge is added) it is probably rare and if it can be decided which case to prefer in terms of performance I think it should be the valid one.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/graph/AbstractBaseGraph.java:
```diff
@@ -293,6 +288,20 @@ public boolean addEdge(V sourceVertex, V targetVertex, E e)
         }
     }
 
+    private boolean addIntrusiveEdge(E e, V sourceVertex, V targetVertex)
```
I prefer to keep two methods with similar code than refactoring to a function here. The reason is that this method reverts changes that happen else where.

HannesWell replied:
My intention was not only to avoid code duplication but also to increase readability/comprehensibility of the code.
But I also understand your argument, so I changed it accordingly.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/graph/BaseIntrusiveEdgesSpecifics.java:
```diff
@@ -148,11 +148,40 @@ public void setEdgeWeight(E e, double weight)
      */
     public abstract boolean add(E e, V sourceVertex, V targetVertex);
 
+    boolean addIntrusiveEdge(E edge, V sourceVertex, V targetVertex, IE e)
```
This needs to be protected. A few of our users have their own specifics implementation and need access here.

HannesWell replied:
I changed the modifier to `protected`.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/graph/BaseIntrusiveEdgesSpecifics.java:
```diff
@@ -148,11 +148,40 @@ public void setEdgeWeight(E e, double weight)
      */
     protected abstract IE getIntrusiveEdge(E e);
+
+    static class IntrusiveEdgeException
```
We also need to give access here.

HannesWell replied:
Done.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/graph/UniformIntrusiveEdgesSpecifics.java:
```diff
@@ -51,24 +51,21 @@ public UniformIntrusiveEdgesSpecifics(Map<E, IntrusiveEdge> map)
 
     @Override
-    public boolean add(E e, V sourceVertex, V targetVertex)
+    public boolean add(E edge, V sourceVertex, V targetVertex)
```
The other methods use `e` for the variable name. No need to change that.

HannesWell replied:
Done.

d-michail reviewed (changes requested):
I prefer to keep this code without an extra 'addIntrusiveEdge` method, even if there is a bit of code duplication.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/graph/BaseIntrusiveEdgesSpecifics.java:
```diff
@@ -148,11 +148,40 @@ public void setEdgeWeight(E e, double weight)
      */
     protected abstract IE getIntrusiveEdge(E e);
+
+    protected static class IntrusiveEdgeException
```
Instead of an inner class, make this a public top-level class.

HannesWell replied:
I changed it accordingly.
But I changed the to arguments to source and target vertex because `IntrusiveEdge` is package-private so users cannot use the class and therefore could not call the constructor.
Or should I make the constructor package-private? If the intention is to only allow users to catch the exception and not to create them, a package-private constructor should be suitable.

By the way the `GraphWalk` class also throws an `InvalidGraphWalkException` that is package-private. So if you want to make the exception catchable by users this should be changed too?

jsichi replied:
Public constructor seems fine to me.

Regarding other existing exceptions, we can consider that in a separate sweeping cleanup issue/pull-request (let's put this one to bed first).  In general, making them explicitly catchable seems like the right thing to do, mostly for helping people write tests and debugging code in cases where they have logic errors in their applications.  Since they're not currently public, I think we can make the changes without having to worry about deprecation.

HannesWell replied:
Good.

Make writing tests and debugging easy for users is definitely a good point. And I think it is outweighed by the opposing desire to keep to keep the API small.
I can open a separate PR to address it once this one is completed.

jsichi commented:
@d-michail if you are good with the latest, can you mark this approved?

jsichi commented:
Thanks!
